src/riak_core_coverage_plan.erl

```diff
@@ -110,120 +293,433 @@ create_plan(VNodeSelector, NVal, PVC, ReqId, Service) ->
             end
     end.
 
+%% @private
+replace_traditional_chunk(VnodeIdx, Node, Filters, NVal,
+                          ReqId, DownNodes, Service, CHBin,
+                          AvailNodeFun) ->
+
+    RingSize = chashbin:num_partitions(CHBin),
+    Offset = ReqId rem NVal,
+    %% We have our own idea of what nodes are available. The client
+    %% may have a different idea of offline nodes based on network
+    %% partitions, so we take that into account.
+    %%
+    %% The client can't really tell us what nodes it thinks are up (it
+    %% only knows hostnames at best) but the opaque coverage chunks it
+    %% uses have node names embedded in them, and it can tell us which
+    %% chunks do *not* work.
+    UpNodes = AvailNodeFun(Service, CHBin) -- [Node|DownNodes],
+    NeededPartitions = partitions_by_index_or_filter(
+                         VnodeIdx, NVal, Filters, RingSize),
+
+    %% For each partition, create a tuple with that partition (as a
+    %% filter index) mapped to a nested tuple of preflist + (initially
+    %% empty) filter list list.
+    Preflists =
+      lists:map(
+        fun(Partition) ->
+                {convert(Partition, keyspace_filter),
+                   safe_hd(
+                     partition_to_preflist(Partition, NVal, Offset, UpNodes, CHBin))
+                }
+        end,
+        NeededPartitions),
+    maybe_create_traditional_replacement(Preflists, lists:keyfind([], 2, Preflists)).
+
+%% @private
+maybe_create_traditional_replacement(Preflists, false) ->
+    %% We do not go to great lengths to minimize the number of
+    %% coverage plan components we'll return, but we do at least sort
+    %% the vnodes we find so that we can consolidate filters later.
+    create_traditional_replacement(lists:sort(Preflists));
+maybe_create_traditional_replacement(_Preflists, _) ->
+    {error, primary_partition_unavailable}.
+
+
+%% @private
+%% Argument to this should be sorted
+create_traditional_replacement(Preflists) ->
+    dechunk_traditional_replacement(
+      lists:foldl(
+        %% Pattern match the current vnode against the head of the
+        %% accumulator; if the vnodes match, we can consolidate
+        %% partition filters
+        fun({PartIdx, VNode}, [{VNode, Partitions}|Tail]) ->
+                [{VNode,
+                  lists:sort([PartIdx|Partitions])}|Tail];
+           %% Instead if this vnode is new, place it into the
+           %% accumulator as is
+           ({PartIdx, VNode}, Accum) ->
+                [{VNode, [PartIdx]}|Accum]
+        end,
+        [],
+        Preflists)).
+
+%% @private
+%% Take our replacement traditional coverage chunks and consolidate it
+%% into a traditional coverage plan (that another layer will rechunk,
+%% but whatchagonnado)
+dechunk_traditional_replacement(Coverage) ->
+    {
+      lists:map(fun({{Vnode, Node}, _Filters}) ->
+                        {Vnode, Node}
+                end, Coverage),
```

Comment:
Torn between asking you to replace this with a list comprehension and leaving as-is.